apps/builder/app/builder/features/ai/operations.tsx

```diff
@@ -0,0 +1,203 @@
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState<string>();
+
+  const jsx = useStore($jsx);
+
+  return (
+    <form
+      method="POST"
+      action=""
+      onSubmit={(event) => {
+        event.preventDefault();
+
+        if (isLoading) {
+          return;
+        }
+
+        const formData = new FormData(event.currentTarget);
+
+        const prompt = formData.get("prompt");
+        let data = formData.get("data");
+
+        if (typeof data === "string") {
+          try {
+            data = JSON.parse(data);
+          } catch {
+            /**/
+          }
+        }
+
+        if (typeof prompt !== "string" || typeof data !== "string") {
+          alert("Invalid data");
+          return;
+        }
+
+        const metas = registeredComponentMetasStore.get();
+        const exclude = ["Body", "Slot"];
+
+        const components = [...metas.keys()].filter(
+          (name) => !exclude.includes(name)
+        );
+
+        abort.current = new AbortController();
+
+        request<operations.Response>(
+          [
+            restAiOperations(),
+            {
+              method: "POST",
+              body: JSON.stringify({
+                prompt,
+                components,
+                jsx: data,
+              }),
+              signal: abort.current.signal,
+            },
+          ],
+          { retry: 2 }
+        )
+          .then((result) => {
+            if (result.success === false) {
+              setError(result.message);
+              return;
+            }
+            applyOperations(result.data);
+          })
+          .finally(() => {
+            abort.current = undefined;
+            setIsLoading(false);
+          });
+
+        setError(undefined);
+        setIsLoading(true);
+      }}
+    >
+      <Label>
+        Prompt
+        <InputField name="prompt" />
+      </Label>
+      <input type="hidden" name="data" value={JSON.stringify(jsx)} />
+      <Button
+        onClick={(event) => {
+          if (isLoading) {
+            event.preventDefault();
+            abort.current?.abort();
+          }
+        }}
+      >
+        {isLoading ? "stop" : "go"}
+      </Button>{" "}
+      {isLoading ? <SpinnerIcon /> : null}
+      {error ? <Text>{error}</Text> : null}
+    </form>
+  );
+};
+
+const $jsx = computed(
+  [
+    selectedInstanceSelectorStore,
+    instancesStore,
+    propsStore,
+    dataSourcesStore,
+    registeredComponentMetasStore,
+    breakpointsStore,
+    stylesStore,
+    styleSourceSelectionsStore,
+  ],
+  (
+    selectedInstanceSelector,
+    instances,
+    props,
+    dataSources,
+    metas,
+    breakpoints,
+    styles,
+    styleSourceSelections
+  ) => {
+    if (selectedInstanceSelector === undefined) {
+      return null;
+    }
+
+    const [rootInstanceId] = selectedInstanceSelector;
+    const instance = instances.get(rootInstanceId);
+    if (instance == null) {
```

Comment:
this makes it look like instance can be null or undefined, I haven't looked, but I think it can only be one of these, we are trying to be more explicit everywhere